### PR TITLE
[All] Remove NOT_FLAG's

### PR DESCRIPTION
**Pull request by rubenzorrilla**
This implements part of the decisions we took in #4896.

As you can see this removes the `NOT_` from the flags. I've substituted all the `Is(NOT_FLAG)` usages by `IsNot(FLAG)`. However, it is not possible to handle the situation in this way in all the cases (e.g. when a `NOT_FLAG` is used as default argument ).

To avoid changing the implementation and behaviour in those cases, I added a new `Not()` static method to the `Flags` class, which is also exported to Python as `__not__()`. This method sets all the bits of the Flag to zeros.

I'm opening a draft PR to have some feedback from the community and to also trigger the CI. 

Finally, I'm tagging the developers who are directly affected by the changes
@loumalouomega (changes in ContactStructuralMechanicsApplication and MeshingApplication)
@ipouplana (changes in PoroechanicsApplication)
@KratosMultiphysics/pfem (changes in DelaunayMeshingApplication)
@AlejandroCornejo (changes in FemToDemApplication)
@miguelmaso (changes in ShallowWaterApplication)

### CODE REVIEW

**Review by loumalouomega — CHANGES_REQUESTED** [review]
I block to prevent accidental merge. In my application should not be a problem, as basically looks like I use it only in the register of flags

**Comment by roigcarlo** [comment]
@KratosMultiphysics/technical-committee agrees on having this as a temporary solution until the `mIsDefined` is completely removed. However, we propose renaming the new `Flags::Not()` method to `Flags::Unset()`.

Please note that this is a **temporary** solution until we implement the second stage discussed in #4896 .

**Comment by rubenzorrilla** [comment]
@KratosMultiphysics/technical-committee the `Not()` method has been renamed to `Unset()` as suggested.
@loumalouomega can you remove your blocking?

**Comment by loumalouomega** [comment]
I need to check once again with the last changes if you don't mind

**Review by loumalouomega — DISMISSED** [review]

**Comment by loumalouomega** [comment]
@rubenzorrilla checked

**Comment by rubenzorrilla** [comment]
I'll wait for the final approval of the @KratosMultiphysics/technical-committee

File: kratos/containers/flags.h
Comment (pooyan-dadvand) [inline kratos/containers/flags.h]:
```suggestion
        Flags unset_this_flag(ThisFlag);
```

**Comment by rubenzorrilla** [comment]
@pooyan-dadvand now I think it's ready

File: applications/DEMApplication/custom_utilities/reorder_consecutive_from_given_ids_model_part_io.h
Comment (pooyan-dadvand) [inline applications/DEMApplication/custom_utilities/reorder_consecutive_from_given_ids_model_part_io.h]: unset DO_NOT_DEFINE_NDEBUG :laughing:

File: applications/DelaunayMeshingApplication/custom_utilities/mesher_utilities.cpp
Comment (pooyan-dadvand) [inline applications/DelaunayMeshingApplication/custom_utilities/mesher_utilities.cpp]:
```suggestion
          pCondition->Set(SELECTED,false); //meaning that is a element that shares faces
```

File: applications/DelaunayMeshingApplication/python_scripts/domain_utilities.py
Comment (pooyan-dadvand) [inline applications/DelaunayMeshingApplication/python_scripts/domain_utilities.py]: why `__unset__` and not `Unset` in python? Is `__unset__` an speciall method?

File: applications/FemToDemApplication/python_scripts/check_and_prepare_model_process.py
Comment (pooyan-dadvand) [inline applications/FemToDemApplication/python_scripts/check_and_prepare_model_process.py]: same

File: kratos/containers/flags.h
Comment (pooyan-dadvand) [inline kratos/containers/flags.h:246]: hmmm, I'm not sure about being a static member. In fact, I am not sure if the behavior is correct. The static one should only mark the define and set the flags to zero. The no static version should unset the given flags bits in the this->mflags and this->mDefined.

File: kratos/processes/entity_erase_process.h
Comment (pooyan-dadvand) [inline kratos/processes/entity_erase_process.h]: For this cases I think we need an EMPTY_FLAGS (with a better name!)

File: kratos/processes/tetrahedral_mesh_orientation_check.h
Comment (pooyan-dadvand) [inline kratos/processes/tetrahedral_mesh_orientation_check.h]: Again an empty flags

File: kratos/processes/tetrahedral_mesh_orientation_check.h
Comment (pooyan-dadvand) [inline kratos/processes/tetrahedral_mesh_orientation_check.h]: same

File: kratos/processes/variational_distance_calculation_process.h
Comment (pooyan-dadvand) [inline kratos/processes/variational_distance_calculation_process.h]: empty flags

File: kratos/processes/variational_distance_calculation_process.h
Comment (pooyan-dadvand) [inline kratos/processes/variational_distance_calculation_process.h]: empty flags

File: kratos/python/add_containers_to_python.cpp
Comment (pooyan-dadvand) [inline kratos/python/add_containers_to_python.cpp]: Is __unset__ an speciall method in python?

File: kratos/tests/cpp_tests/containers/test_flags.cpp
Comment (pooyan-dadvand) [inline kratos/tests/cpp_tests/containers/test_flags.cpp]: I think ~ is the correct operator instead of Unset... Isn't it?

**Review by pooyan-dadvand — CHANGES_REQUESTED** [review]
I think we should give this another iteration. Hopefuly the last one! :smirk:

**Comment by pooyan-dadvand** [comment]
After giving it another try we (@KratosMultiphysics/technical-committee ) arrived to following options for passing a former NOT_ flags to the constructor considering the following example:
```c++
ModelPartIO(std::string const& Filename, const Flags Options = IO::READ|IO::NOT_IGNORE_VARIABLES_ERROR|IO::SKIP_TIMER);
```
1. Having an static `SetToFalse` method of `Flag`
```c++
ModelPartIO(std::string const& Filename,  const Flags Options = 
    IO::READ|
    Flags::SetToFalse(IO::IGNORE_VARIABLES_ERROR)  |
    IO::SKIP_TIMER);
```
2. Having an static `False` method of `Flag`
```c++
ModelPartIO(std::string const& Filename, const Flags Options = 
    IO::READ|
    Flags::False(IO::IGNORE_VARIABLES_ERROR)|
    IO::SKIP_TIMER);
```
3. Having a `False` method of `Flags`
```c++
ModelPartIO(std::string const& Filename, const Flags Options = 
    IO::READ|
    IO::IGNORE_VARIABLES_ERROR.False()|
    IO::SKIP_TIMER);
```
Our favorite is the last option but we are open to other suggestions and coments.

**Comment by pooyan-dadvand** [comment]
As there are not comments neither suggestions, @KratosMultiphysics/technical-committee agrees on the third option.

File: applications/DelaunayMeshingApplication/python_scripts/domain_utilities.py
Comment (rubenzorrilla) [reply applications/DelaunayMeshingApplication/python_scripts/domain_utilities.py]: I don't know :sweat_smile: . The false will be `False()` in Python too.

File: kratos/tests/cpp_tests/containers/test_flags.cpp
Comment (rubenzorrilla) [reply kratos/tests/cpp_tests/containers/test_flags.cpp]: Using the `~` doesn't yield the same result (it's the first thing I tested).

**Comment by rubenzorrilla** [comment]
> As there are not comments neither suggestions, @KratosMultiphysics/technical-committee agrees on the third option.

I found out that the Python `False` is also a keyword when naming functions and classes. Hence, I propose to go ahead with the 1st option (`SetToFalse` method of `Flags`).

**Comment by pooyan-dadvand** [comment]
@KratosMultiphysics/technical-committee suggests to change the name of the method from `False` to `AsFalse` to be the same in Python and cpp without having syntax problem in Python.

**Comment by rubenzorrilla** [comment]
@pooyan-dadvand I think that now it is ready

File: applications/ContactStructuralMechanicsApplication/custom_processes/find_intersected_geometrical_objects_with_obb_for_contact_search_process.h
Comment (philbucher) [inline applications/ContactStructuralMechanicsApplication/custom_processes/find_intersected_geometrical_objects_with_obb_for_contact_search_process.h]: shouldn't this be sth with `AsFalse`?
```suggestion
            BaseType::DEBUG_OBB.AsFalse()|
```
Not sure though

File: kratos/mpi/tests/sources/test_flag_synchronization.cpp
Comment (philbucher) [inline kratos/mpi/tests/sources/test_flag_synchronization.cpp:148]: did you run those tests also in MPI?

**Review by philbucher — COMMENTED** [review]
minor comments otherwise fine to me :+1:

File: kratos/mpi/tests/sources/test_flag_synchronization.cpp
Comment (rubenzorrilla) [reply kratos/mpi/tests/sources/test_flag_synchronization.cpp:148]: No. In any case these changes do not affect the parallelism.

Comment (philbucher) [reply kratos/mpi/tests/sources/test_flag_synchronization.cpp:148]: Hm I wouldn't be so sure, there are functions to sync the flags, which if I remember correctly somehow use the flags and its internals
can you please run the tests?

File: applications/ContactStructuralMechanicsApplication/custom_processes/find_intersected_geometrical_objects_with_obb_for_contact_search_process.h
Comment (philbucher) [inline applications/ContactStructuralMechanicsApplication/custom_processes/find_intersected_geometrical_objects_with_obb_for_contact_search_process.h:110]: @rubenzorrilla why did this even compile?
Shouldn't this operator be removed?
Or is it still used somewhere else?

Comment (rubenzorrilla) [reply applications/ContactStructuralMechanicsApplication/custom_processes/find_intersected_geometrical_objects_with_obb_for_contact_search_process.h:110]: The point is that the operator `~` is not the same as our flavour of the `False`. This is why we keep it.

File: kratos/mpi/tests/sources/test_flag_synchronization.cpp
Comment (rubenzorrilla) [reply kratos/mpi/tests/sources/test_flag_synchronization.cpp:148]: I've just run the MPI suite. Everything works as expected :wink:

Comment (philbucher) [reply kratos/mpi/tests/sources/test_flag_synchronization.cpp:148]: Ok thx
No worries, I am anyway planning to soon add the MPI tests to the CI so you will be blocked from merging if you screw things up :)))

**Review by philbucher — APPROVED** [review]
Ping @pooyan-dadvand

**Comment by roigcarlo** [comment]
Dear @KratosMultiphysics/all Please be aware that this is due to merge next week. (02/03/2020)

**Review by pooyan-dadvand — APPROVED** [review]

**Pull request merged**
